pkg/providers/vmware/provider.go

```diff
@@ -0,0 +1,422 @@
+	if err != nil {
+		return "", err
+	}
+	return vm.Name(), nil
+}
+
+// GetVMStatus gets the power status of the source VM
+func (r *VmwareProvider) GetVMStatus() (provider.VMStatus, error) {
+	vmProperties, err := r.getVmProperties()
+	if err != nil {
+		return "", err
+	}
+
+	poweredOn := types.VirtualMachinePowerStatePoweredOn
+	poweredOff := types.VirtualMachinePowerStatePoweredOff
+
+	switch vmProperties.Runtime.PowerState {
+	case poweredOn:
+		return provider.VMStatusUp, nil
+	case poweredOff:
+		return provider.VMStatusDown, nil
+	}
+
+	return "", fmt.Errorf("VM doesn't have a legal status. Allowed statuses: [%v, %v]", poweredOn, poweredOff)
+}
+
+// StartVM powers on the source VM.
+func (r *VmwareProvider) StartVM() error {
+	vmwareClient, err := r.getClient()
+	if err != nil {
+		return err
+	}
+	vm, err := r.getVM()
+	if err != nil {
+		return err
+	}
+	return vmwareClient.StartVM(vm.Reference().Value)
+}
+
+// StartVM powers off the source VM.
+func (r *VmwareProvider) StopVM(cr *v1beta1.VirtualMachineImport, client client.Client) error {
+	vmwareClient, err := r.getClient()
+	if err != nil {
+		return err
+	}
+	vm, err := r.getVM()
+	if err != nil {
+		return err
+	}
+	err = vmwareClient.StopVM(vm.Reference().Value)
+	if err != nil {
+		return err
+	}
+	err = utils.AddFinalizer(cr, utils.RestoreVMStateFinalizer, client)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CleanUp removes transient resources created for import
+func (r *VmwareProvider) CleanUp(failure bool, cr *v1beta1.VirtualMachineImport, client client.Client) error {
+	var errs []error
+
+	err := utils.RemoveFinalizer(cr, utils.RestoreVMStateFinalizer, client)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	vmiName := k8stypes.NamespacedName{
+		Name:      r.vmiObjectMeta.Name,
+		Namespace: r.vmiObjectMeta.Namespace,
+	}
+
+	err = r.secretsManager.DeleteFor(vmiName)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	err = r.configMapsManager.DeleteFor(vmiName)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	if failure {
+		err = r.dataVolumesManager.DeleteFor(vmiName)
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		err = r.virtualMachineManager.DeleteFor(vmiName)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return foldErrors(errs, vmiName)
+	}
+	return nil
+}
+
+// TestConnection tests the connection to the vCenter or ESXi host.
+func (r *VmwareProvider) TestConnection() error {
+	vmwareClient, err := r.getClient()
+	if err != nil {
+		return err
+	}
+	return vmwareClient.TestConnection()
+}
+
+// Validate checks whether the source VM and resource mapping is valid.
+func (r *VmwareProvider) Validate() ([]v1beta1.VirtualMachineImportCondition, error) {
+	// TODO: implement vmware rule validation
+	return []v1beta1.VirtualMachineImportCondition{
+		conditions.NewCondition(v1beta1.Valid, string(v1beta1.ValidationCompleted), "Validation completed successfully", corev1.ConditionTrue),
```

Comment:
That method needs to set both `v1beta1.Valid` and `v1beta1.MappingRulesVerified` conditions, where `Valid` is responsible for resource mapping checks and `MappingRulesVerified` for source VM checks.